### PR TITLE
[#1529] feat(spark-connector): support table operation

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
1. support basic table DDL and DML operation.  (alter table are not supported)
2. add basic table operation to GravitinoCatalog
3. add GravitinoBaseTable class which get schema from Gravitino and do IO with internal spark catalog table.

### Why are the changes needed?

Fix: #1529 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
add spark SQL IT

### CODE REVIEW

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -45,6 +45,7 @@ postgresql = "42.6.0"
 immutables-value = "2.10.0"
 selenium = "3.141.59"
 rauschig = "1.2.0"
+kyuubi = "1.8.0"
```
Comment (pan3793) [inline gradle/libs.versions.toml:48]: Kyuubi 1.8.1 is under RC[1] period, which provides Scala 2.13 jars in the Maven repo, you may want to have a try with the staging repo[2].

[1] https://www.mail-archive.com/dev@kyuubi.apache.org/msg03096.html
[2] https://repository.apache.org/content/repositories/orgapachekyuubi-1042/

Comment (FANNG1) [reply gradle/libs.versions.toml:48]: got it, thanks!

**Comment by FANNG1** [comment]
@jerryshao @qqqttt123 @yuqi1129 @diqiu50 ,  it's ready to review now, please help to review when you are free.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    public static SparkColumnInfo of(String name, String type) {
+      return of(name, type, null);
+    }
+
+    public static SparkColumnInfo of(String name, String type, String comment) {
+      return new SparkColumnInfo(name, type, comment);
+    }
+  }
+
+  public enum ParseStage {
```
Comment (qqqttt123) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:64]: What's the meaning of `ParseStage`? Is it a standard term?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    COLLUMN,
+    DETECT,
+    PARTITION,
+    PARTITIONBUCKET,
+    TABLE_INFO,
+  }
+
+  public static SparkTableInfo getSparkTableInfo(List<Object[]> rows) {
```
Comment (qqqttt123) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:72]: Let me think twice.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    ParseStage stage = ParseStage.COLLUMN;
+    SparkTableInfo tableInfo = new SparkTableInfo();
+    for (Object[] os : rows) {
```
Comment (qqqttt123) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:75]: Could use  org.apache.spark.sql.connector.catalog.Table.properties and partitioning?
If you use Hive, maybe you can use Hive client to get table information, too.

Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:75]: I prefer to parse spark sql result:
1. it's general and could reused by other catalogs like Iceberg and JDBC
2. it's necessary to check Spark SQL output

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java (hunk 2 above)
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:64]: I use something like state machine to parse SparkSQL result,  `ParseStage` is used to check which state or stage is current on.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java (hunk 4 above)
Comment (qqqttt123) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:75]: It has bad compatible issues. If Spark community change the output, it will bring issues to you.

Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:75]: yes, seems hard to get spark table meta

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/GravitinoBaseTable.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.StructType$;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+/**
+ * Provides schema info from Gravitino, IO from the internal spark table. The specific catalog table
+ * could implement more capabilities like SupportsPartitionManagement for Hive table, SupportsIndex
+ * for JDBC table, SupportsRowLevelOperations for Iceberg table.
+ */
+public class GravitinoBaseTable implements Table, SupportsRead, SupportsWrite {
```
Comment (jerryshao) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/table/GravitinoBaseTable.java:39]: Here is it better renaming to `SparkBaseTable`?

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/SparkTypeConverter.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+public class SparkTypeConverter {
+  public static Type toGravitinoType(DataType sparkType) {
+    if (sparkType instanceof StringType) {
+      return Types.StringType.get();
+    } else if (sparkType instanceof BooleanType) {
+      return Types.BooleanType.get();
+    } else if (sparkType instanceof IntegerType) {
+      return Types.IntegerType.get();
+    }
+    throw new UnsupportedOperationException("Not support " + sparkType.toString());
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/SparkTypeConverter.java:28]: You mean to support more spark types or more spark version?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+          case "Comment":
+            tableInfo.comment = items[1];
+            break;
+          default:
+            tableInfo.addUnknownItem(String.join("@", items));
+        }
+      }
+    }
+    return tableInfo;
+  }
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:129]: > I'm a bit confused of what you're doing here, are you parsing the table information from an output string? So basically:
> 
> 1. What is the usage of this method?

parse table information from Spark SQL to a structured SparkTableInfo, facility to do the result check.

> 2. Why do you rely on the output string, it is quite hacky and fragile.

Spark doesn't provide a API to get table meta,  I couldn't find a general way too. Spark use internal API to do some tests and Iceberg use Iceberg Catalog API to verify the result. 

> 3. What if user uses dataframe, temp view, can you still get such information?

I think we could parse `temp view` output it's like table. for dataframe we could use a specific parser or just do string compare.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfoChecker.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+  public static SparkTableInfoChecker create() {
+    return new SparkTableInfoChecker();
+  }
+
+  public enum CheckField {
+    NAME,
+    COLUMN,
+    COMMENT,
+  }
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfoChecker.java:31]: make it private

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java [hunk 9]
```diff
@@ -46,6 +48,27 @@ protected Set<String> getDatabases() {
...
+  protected void dropDatabaseIfExists(String database) {
+    sql("DROP DATABASE IF EXISTS " + database);
+  }
+
+  // Specify Location explicitly because the default location is local HDFS, Spark will expand the
+  // location to HDFS.
+  protected void createDatabaseIfNotExists(String database) {
+    sql(
+        String.format(
+            "create database IF NOT EXISTS %s LOCATION '/user/hive/%s'", database, database));
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java]: done

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/GravitinoCatalogAdaptor.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+   * Create a specific Spark table, combined with gravitinoTable to do DML operations and
+   * sparkCatalog to do IO operations.
+   *
+   * @param identifier Spark's table identifier
+   * @param gravitinoTable Gravitino table to do DDL operations
+   * @param sparkCatalog specific Spark catalog to do IO operations
+   * @param propertiesConverter transform properties between Gravitino and Spark @Retrun a specific
+   *     Spark table
+   */
+  GravitinoBaseTable createSparkTable(
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/GravitinoCatalogAdaptor.java]: done

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+  // TO: will implement it in property PR.
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java]: done

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 12]
```diff
@@ -50,32 +52,99 @@ public class GravitinoCatalog implements TableCatalog, SupportsNamespaces {
...
+    NameIdentifier nameIdentifier =
+        NameIdentifier.of(metalakeName, catalogName, getDatabase(ident), ident.name());
+    com.datastrato.gravitino.rel.Column[] gravitinoColumns =
+        Arrays.stream(columns)
+            .map(column -> createGravitinoColumn(column))
+            .toArray(com.datastrato.gravitino.rel.Column[]::new);
+
+    Map<String, String> gravitinoProperties =
+        propertiesConverter.toGravitinoTableProperties(properties);
+    String comment = gravitinoProperties.remove("comment");
```
Comment (diqiu50) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:119]: Removing this attribute and retrieving the 'comment' is too abrupt an implementation. can you give an easy to understand implementation

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 13]
```diff
@@ -88,32 +157,45 @@ public Table createTable(
...
 
   @Override
-  public void renameTable(Identifier oldIdent, Identifier newIdent)
-      throws NoSuchTableException, TableAlreadyExistsException {
-    throw new NotSupportedException("Doesn't support renaming table");
+  public boolean purgeTable(Identifier ident) {
+    checkNotPathIdentifier(ident, "purgeTable");
+    return gravitinoCatalogClient
+        .asTableCatalog()
+        .purgeTable(NameIdentifier.of(metalakeName, catalogName, getDatabase(ident), ident.name()));
```
Comment (diqiu50) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:176]: Missing `Exception` Handling, the drop table too

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    if (gravitinoTable.properties() != null) {
+      properties.putAll(gravitinoTable.properties());
+    }
+
+    properties = propertiesConverter.toSparkTableProperties(properties);
+
+    String comment = gravitinoTable.comment();
+    if (comment != null) {
+      properties.put(ConnectorConstants.COMMENT, comment);
+    }
```
Comment (diqiu50) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:110]: It's best to handle comments together.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 13 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:176]: Should I process `RuntimeException` explicitly?

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 12 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:119]: Unlike trino store comment in table meta data, spark stores comment in properties, seems it's necessary to do the transform. any suggestions?

Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java]: add some description to related code.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java (hunk 14 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:111]: add some descriptions

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jerryshao) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:129]: Can we do the similar way like Iceberg?

[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkTableInfo.java:129]: there are some drawbacks about Iceberg way,   we could test create and alter operation with Gravitino catalog but couldn't test the load operations like loadTable

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 12 above)
Comment (diqiu50) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java]: My suggestion is to first separate the handling of 'getComment' and 'remove unused attribute' for easier understanding. Secondly, extracting 'comment' from attributes and setting 'comment' into attributes should be managed within a single class.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 13 above)
Comment (diqiu50) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:178]: You can refer to the exception handling in a module similar to Spark.

Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:178]: I will check it latter

**Comment by FANNG1** [comment]
We could get `ResolvedTable` from the dataset. ResolvedTable contains `SparkBaseTable` returned by GravitinoCatalog with following interfaces. If we test with SparkBaseTable, we could get a structured table metadata but not 100% end2end, such as:  `comment`, `location` is hidden in properties when testing table properties,  we could only check an expected transform object when testing partition. do you think it's a better way than parsing spark SQL result? @jerryshao @qqqttt123 

```
public interface Table {

  /**
   * A name to identify this table. Implementations should provide a meaningful name, like the
   * database and table name from catalog, or the location of files for this table.
   */
  String name();

  /**
   * Returns the schema of this table. If the table is not readable and doesn't have a schema, an
   * empty schema can be returned here.
   * <p>
   * This is deprecated. Please override {@link #columns} instead.
   */
  @Deprecated
  StructType schema();

  /**
   * Returns the columns of this table. If the table is not readable and doesn't have a schema, an
   * empty array can be returned here.
   */
  default Column[] columns() {
    return CatalogV2Util.structTypeToV2Columns(schema());
  }

  /**
   * Returns the physical partitioning of this table.
   */
  default Transform[] partitioning() {
    return new Transform[0];
  }

  /**
   * Returns the string map of table properties.
   */
  default Map<String, String> properties() {
    return Collections.emptyMap();
  }

  /**
   * Returns the set of capabilities for this table.
   */
  Set<TableCapability> capabilities();
}
```

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 13 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:178]: seems no need to handle exception like trino, `dropTable` return false means table not exists, Spark will check the exists of the table before `dropTable`
```
    boolean dropped =
        tableCatalog.dropTable(
            NameIdentifier.ofTable(
                metalake.name(), catalogName, tableName.getSchemaName(), tableName.getTableName()));
    if (!dropped) {
      throw new TrinoException(GRAVITINO_OPERATION_FAILED, "Failed to drop table " + tableName);
    }
```

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java [hunk 15]
```diff
@@ -78,8 +83,14 @@ protected List<Object[]> sql(String query) {
...
   }
 
+  // It's hard to parse Spark SQL output, create SparkTableInfo from SparkBaseTable.
   protected SparkTableInfo getTableInfo(String tableName) {
-    return SparkTableInfo.getSparkTableInfo(sql("DESC TABLE EXTENDED " + tableName));
+    Dataset ds = getSparkSession().sql("DESC TABLE EXTENDED " + tableName);
+    CommandResult result = (CommandResult) ds.logicalPlan();
+    DescribeRelation relation = (DescribeRelation) result.commandLogicalPlan();
+    ResolvedTable table = (ResolvedTable) relation.child();
+    SparkBaseTable baseTable = (SparkBaseTable) table.table();
```
Comment (diqiu50) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java:92]: Can you confirm that the force type conversion was successful?

Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java:92]: For current spark version, I think it's YES, For other spark versions may need some adaptor work.

**Comment by jerryshao** [comment]
@FANNG1 can you please fix the conflicts.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    this.identifier = identifier;
+    this.gravitinoTable = gravitinoTable;
+    this.sparkCatalog = sparkCatalog;
+    this.propertiesConverter = propertiesConverter;
+  }
+
+  protected Table getSparkTable() {
+    if (lazySparkTable == null) {
+      try {
+        lazySparkTable = this.sparkCatalog.loadTable(identifier);
```
Comment (jerryshao) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:60]: `this.lazySparkTable = sparkCatalog.xxx`

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+                            .build();
+                  }
+                  return StructField.apply(
+                      column.name(),
+                      SparkTypeConverter.toSparkType(column.dataType()),
+                      column.nullable(),
+                      metadata);
+                })
+            .collect(Collectors.toList());
+    return StructType$.MODULE$.apply(structs);
```
Comment (jerryshao) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:95]: Do yo support nest fields here?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 18]
```diff
@@ -74,4 +113,192 @@ void testDropSchema() {
...
+    sql("USE " + databaseName);
+    createSimpleTable(tableName);
+    tableInfo = getTableInfo(tableName);
+    checker =
+        SparkTableInfoChecker.create().withName(tableName).withColumns(getSimpleTableColumn());
+    checker.check(tableInfo);
+    checkTableReadWrite(tableInfo);
+  }
+
+  // Create tables with advanced features like comment, property, partition, bucket etc.
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:167]: will implement in another PR, will modify the description

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java [hunk 19]
```diff
@@ -55,6 +83,33 @@ protected List<Object[]> sql(String query) {
     return rowsToJava(rows);
   }
 
+  // It's hard to parse Spark SQL output, create SparkTableInfo from SparkBaseTable.
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java:86]: ok

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java (hunk 17 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:95]: no, only support Integer, bool, String, more types will be supported in another PR

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 20]
```diff
@@ -74,4 +113,192 @@ void testDropSchema() {
     Assertions.assertThrowsExactly(
         NoSuchNamespaceException.class, () -> sql("drop database notExists"));
   }
+
+  // Create simple tables without advanced features like comment, property, partition, bucket.
+  // To check whether create and load operation with default.
+  @Test
+  void testCreateSimpleTable() {
+    String tableName = "simpleTable";
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:121]: by default spark is not case-sensitive,   let me check the overall logic to fix.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java [hunk 21]
```diff
@@ -55,6 +83,33 @@ protected List<Object[]> sql(String query) {
...
+    try {
+      SparkTableInfo tableInfo = getTableInfo(tableName);
+      Assertions.assertEquals(tableName, tableInfo.getTableName());
+      return true;
+    } catch (Exception e) {
+      if (e instanceof AnalysisException) {
+        return false;
+      }
+      throw e;
+    }
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java:110]: seems couldn't use [Throwables](https://guava.dev/releases/23.0/api/docs/com/google/common/base/Throwables.html) to simplify the code, `Throwables` could transform one exception to another. but current logic we should do specific handle for AnalysisException.  As AnalysisException is not declared, we couldn't simplify code like this too:
```java
} catch (AnalysisException e) {
    return false;
} catch (Exception e) {
    throw e;
}
```

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java (hunk 16 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java]: done

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 22]
```diff
@@ -50,32 +53,100 @@ public class GravitinoCatalog implements TableCatalog, SupportsNamespaces {
...
+    if (namespace.length == 0) {
+      gravitinoNamespace = getCatalogDefaultNamespace();
+    } else {
+      validateNamespace(namespace);
+      gravitinoNamespace = namespace[0];
+    }
+    NameIdentifier[] identifiers =
+        gravitinoCatalogClient
+            .asTableCatalog()
+            .listTables(Namespace.of(metalakeName, catalogName, gravitinoNamespace));
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java]: fixed and add test

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java]: fixed

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java [hunk 23]
```diff
@@ -74,4 +113,192 @@ void testDropSchema() {
...
+            .withName(tableName)
+            .withColumns(getSimpleTableColumn())
+            .withComment(null);
+    checker.check(tableInfo);
+
+    checkTableReadWrite(tableInfo);
+  }
+
+  // testCreateSimpleTable create tables without database name,
+  // while this test create tables with specific database name.
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java]: fixed

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:120]: SparkSQL itself doesn't care case-sensitive, but the underlying catalog cares. Hive is not case-sensitive, JDBC is case-sensitive. For spark-connector I think there's no need to handle case-sensitive issues. just need to pass through the name from the underlying catalog to spark SQL.  do you agree?

[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:120]: For this cases, spark-connector pass origin name not the name from Gravitino back to SparkSQL, so the IT passes. I think we should pass the name from Gravitino to SparkSQL.

[the remark this answers is not in the document]
Comment (jerryshao) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java:120]: So does here really test from end to end, what will the actual table be created in Gravitino? If it is a hive table, does it really support table name like "simpleTable"?

**Comment by qqqttt123** [comment]
The package should be consistent with the trino.
`com.datastrato.gravitino.spark.connector`

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+
+  protected Table getSparkTable() {
+    if (this.lazySparkTable == null) {
+      try {
+        this.lazySparkTable = sparkCatalog.loadTable(identifier);
+      } catch (NoSuchTableException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    return this.lazySparkTable;
```
Comment (jerryshao) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java:65]: Adding `this.` only when the class variable is assigned/initialized.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 25]
```diff
@@ -191,41 +279,50 @@ public boolean dropNamespace(String[] namespace, boolean cascade)
...
+        "Only support 3 level namespace," + gravitinoIdentifier.namespace());
+    return new String[] {gravitinoIdentifier.namespace().level(2)};
   }
 
-  private void validateNamespace(String[] namespace) {
+  private static void checkNotPathIdentifier(Identifier identifier, String method) {
     Preconditions.checkArgument(
-        namespace.length == 1,
-        "Doesn't support multi level namespaces: " + String.join(".", namespace));
+        !(identifier instanceof PathIdentifier),
```
Comment (qqqttt123) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:324]: It is  a little weird that we depend the Iceberg here.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 26]
```diff
@@ -50,32 +53,104 @@ public class GravitinoCatalog implements TableCatalog, SupportsNamespaces {
...
   public Table loadTable(Identifier ident) throws NoSuchTableException {
-    throw new NotSupportedException("Doesn't support loading table");
+    checkNotPathIdentifier(ident, "loadTable");
+    try {
+      com.datastrato.gravitino.rel.Table table =
+          gravitinoCatalogClient
+              .asTableCatalog()
+              .loadTable(
+                  NameIdentifier.of(metalakeName, catalogName, getDatabase(ident), ident.name()));
+      return gravitinoAdaptor.createSparkTable(ident, table, sparkCatalog, propertiesConverter);
```
Comment (qqqttt123) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:150]: Why do the Adaptor call createSparkTable insteaf of `loadSparkTable` here?

**Comment by FANNG1** [comment]
> The package should be consistent with the trino. `com.datastrato.gravitino.spark.connector`

ok

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 26 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:150]: `loadSparkTable` sounds like fetch a remote table to local, while `createSparkTable` just create a SparkTable object.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 25 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:324]: will remove it.

**Comment by FANNG1** [comment]
> The package should be consistent with the trino. `com.datastrato.gravitino.spark.connector`

I will propose a separate PR to fix this, see #2442

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java (hunk 24 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java]: done

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java (hunk 25 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java]: done

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkIT.java]: yes, the test is from end to end,  take create hive table SimpleTable for example,
1. spark SQL pass create table request to spark-connector, with identifier `SimpleTable`
2. spark-connector request create table to Gravitino with identifier `SimpleTable`
3. Gravitino creates hive table with identifier  `SimpleTable`
4. Hive metastore create table with `simpletable`, because it's not case sensitive
Load  table process:
6. Spark SQL load table `SimpleTable` ,  spark connector request Gravitino with this identifier too.
7.  The identifier returned by Gravitino is `simpletable`, but is not used by spark-connector
8. spark connector pass origin `SimpleTable` to SparkSQL, so the test check passes.

we should pass the identifier returned by Gravitino to keep consistent with SparkSQL origin actions.

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java [hunk 27]
```diff
@@ -52,32 +54,104 @@ public class GravitinoCatalog implements TableCatalog, SupportsNamespaces {
...
+    } catch (NoSuchSchemaException e) {
+      throw new NoSuchNamespaceException(ident.namespace());
+    } catch (com.datastrato.gravitino.exceptions.TableAlreadyExistsException e) {
+      throw new TableAlreadyExistsException(ident);
+    }
   }
 
-  // Will create a catalog specific table by invoking createSparkTable()
+  // Will create a catalog specific table
```
Comment (jerryshao) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:142]: Is this comment valid?

Comment (jerryshao) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java:142]: If you're specifying line 151, probably you need to add comment to that line.

Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/catalog/GravitinoCatalog.java]: moved to 151

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.StructType$;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+
+/**
+ * Provides schema info from Gravitino, IO from the internal spark table. The specific catalog table
+ * could implement more capabilities like SupportsPartitionManagement for Hive table, SupportsIndex
+ * for JDBC table, SupportsRowLevelOperations for Iceberg table.
+ */
+public class SparkBaseTable implements Table, SupportsRead, SupportsWrite {
```
[the remark this answers is not in the document]
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/table/SparkBaseTable.java]: add abstract

**Review by jerryshao — APPROVED** [review]
LGTM.
